File: kafka/fake-data-pipeline/projeto/consumer.py
from kafka import KafkaConsumer
from kafka import KafkaProducer,KafkaConsumer
from json import loads
import os

import findspark
findspark.init()

from pyspark.sql import SparkSession

spark = SparkSession.builder.appName("ArCondicionado").getOrCreate()

df = spark \
    .readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", "kafka-single-node_kafka_1:29092") \
    .option("subscribe", "devices") \
    .load()
    
df.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")


producer = KafkaProducer(bootstrap_servers='kafka-single-node_kafka_1:29092',
                        value_serializer=lambda v:str(v).encode('utf-8'))


consumer = KafkaConsumer(
    'kafka-python-topic',
     bootstrap_servers=['kafka-single-node_kafka_1:29092'],
     auto_offset_reset='earliest',
     enable_auto_commit=True,
     value_deserializer=lambda x: loads(x.decode('utf-8')))

def liga_ar(avg):
    if avg >= 30 : 
        flag = 1
    else : 
        flag = 0
    if flag == 1 :
        print('\nLigar o ar condicionado !')
    return flag

from statistics import mean
list10 = []
for msg in consumer:
    print(msg)
    message = msg.value
    print(message)
    while len (list10) > 10 :
        list10.pop(0)
        media = mean(list10)
        flag = liga_ar(media)
        producer.send('devices', flag)
        print("Média: ",media)
    list10.append(message)

File: kafka/fake-data-pipeline/projeto/Consumer.scala
import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.streaming._
import org.apache.spark.sql.types._

object Consumer {
  def main(args: Array[String]): Unit = {
    //import java.util.Properties
    //
    //// properties for jdbc
    //val properties = new Properties()
    //properties.put("user", "postgres")
    //properties.put("password", "**************")

    val spark = SparkSession
      .builder
      .master("yarn")
      .appName("Consumer")
      .getOrCreate()

    import spark.implicits._

    val inputDF = spark.readStream
      .format("kafka")
      .options(Map(
        "kafka.bootstrap.servers" -> "kafka-single-node:9092",
        "subscribe" -> "kafka-scala-topic")
      )
      .load()
    
    val rawDF = inputDF.selectExpr("topic", "CAST(value AS STRING)", "timestamp")
    //print(rawDF)
    val query = rawDF.writeStream
      .trigger(Trigger.ProcessingTime("10 seconds"))
      .foreachBatch { (batchDF: DataFrame, _) =>
        // cache
        batchDF.persist()
        //println(batchDF.show())
        
        // Topic: iot
        batchDF.where($"topic" === "kafka-scala-topic")
          .withColumn("_tmp",$"value") 
          .agg(avg("_tmp"))
          .withColumnRenamed("avg(_tmp)", "temp")
          .show(3)
    
        println("Value from kafka")
        //println(batchDF.show())
        //println("write to postgresql")

        // uncache
        batchDF.unpersist()
      }
      .outputMode("update")
      .start()

    query.awaitTermination()
  }


}
